### ISSUE TITLE
Dataset not ending when created inside predict_step function.

### ISSUE THREAD

**Issue by ramonemiliani93** [tensorflow/tensorflow#47534]
<em>Please make sure that this is a bug. As per our
[GitHub Policy](https://github.com/tensorflow/tensorflow/blob/master/ISSUES.md),
we only address code/doc bugs, performance issues, feature requests and
build/installation issues on GitHub. tag:bug_template</em>

**System information**
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): MacOS Catalina 10.15.6
- TensorFlow installed from (source or binary): source
- TensorFlow version (use command below): 2.4.1
- Python version: 3.7.9 

**Describe the current behavior**
When overriding the `predict_step` of a `tf.keras.Model` and creating a `Dataset` inside, it will not end.

**Describe the expected behavior**
`Dataset` should end.

**Standalone code to reproduce the issue**
https://colab.research.google.com/drive/1lKF1KxUdUDjJq_DTfle5HZITWIcICgdG?usp=sharing

**Comment by amahendrakar** [comment]
Was able to reproduce the issue with TF v2.3, TF v2.4 and TF-nightly. Please find the gist of it [here](https://colab.research.google.com/gist/amahendrakar/3103c98f5f9947333fe79ce72631b59e/47534.ipynb). Thanks!

**Comment by ramonemiliani93** [comment]
Seems the problem is on the `TensorSliceDataset` class specifically where `self._tensors` attribute is created. Seems `structure.to_batched_tensor_list` is not generating the correct number of tensors (only 1).

**Comment by ramonemiliani93** [comment]
The `nest.flatten` function inside `normalize_elements` to generate `element` on the `__init__` of `TensorSliceDataset` works different when inside the `predict_step`.

**Comment by ramonemiliani93** [comment]
The problem seems to be on the `nest.flatten` inside `normalize_element` when creating `element` in `TensorSliceDataset`, for a simple example i.e. `tf.data.Dataset.from_tensor_slices(tf.random.uniform((120, 1, 5))` a list is generated correctly. Inside the `predict_step` a list with a single element is created (instead of unpacking all tensors on the first dimension).

**Comment by sushreebarsa** [comment]
Was able to reproduce the issue in TF v2.5,please find the gist [here](https://colab.research.google.com/gist/sushreebarsa/feceeca87ab1701a9850f517d877b149/untitled144.ipynb)..Thanks !
